mlir/lib/Dialect/Vector/Utils/VectorUtils.cpp

```diff
@@ -300,3 +301,132 @@ vector::createUnrollIterator(VectorType vType, int64_t targetRank) {
   shapeToUnroll = shapeToUnroll.slice(0, firstScalableDim);
   return StaticTileOffsetRange(shapeToUnroll, /*unrollStep=*/1);
 }
+
+FailureOr<vector::ConstantOrScalableBound::BoundSize>
+vector::ConstantOrScalableBound::getSize() const {
+  if (map.isSingleConstant())
+    return BoundSize{map.getSingleConstantResult(), /*scalable=*/false};
+  if (map.getNumResults() != 1 || map.getNumInputs() != 1)
+    return failure();
+  auto binop = dyn_cast<AffineBinaryOpExpr>(map.getResult(0));
+  if (!binop || binop.getKind() != AffineExprKind::Mul)
+    return failure();
+  auto matchConstant = [&](AffineExpr expr, int64_t &constant) -> bool {
+    if (auto cst = dyn_cast<AffineConstantExpr>(expr)) {
+      constant = cst.getValue();
+      return true;
+    }
+    return false;
+  };
+  // Match `s0 * cst` or `cst * s0`:
+  int64_t cst = 0;
+  auto lhs = binop.getLHS();
+  auto rhs = binop.getRHS();
+  if ((matchConstant(lhs, cst) && isa<AffineSymbolExpr>(rhs)) ||
+      (matchConstant(rhs, cst) && isa<AffineSymbolExpr>(lhs))) {
+    return BoundSize{cst, /*scalable=*/true};
+  }
+  return failure();
+}
+
+namespace {
+struct ScalableValueBoundsConstraintSet : public ValueBoundsConstraintSet {
```

Comment:
I'm trying to see if we can reuse more functionality from `ValueBoundsConstraintSet`, so less has to be reimplemented here.

One thing that stands out to me in this PR is that `vector.vscale` does not implement `ValueBoundsOpInterface`.

Here's an idea that could work:
- `ScalableValueBoundsConstraintSet` defines 3 fields: `vscaleMin`, `vscaleMax` and `vscale`.
- `vscaleMin` and `vscaleMax` are initialized in the constructor of `ScalableValueBoundsConstraintSet`.
- `vscale` is first seen `vscale` SSA value. "empty" value in the beginning.
- The `ValueBoundsOpInterface::populateBoundsForIndexValue` implementation dyn_casts `cstr` to `ScalableValueBoundsConstraintSet &`. If that succeeds, we can get `vscaleMin`, `vscaleMax` and `vscale` and implement the same logic that you have here without the `PopulateCustomValueBoundsFn` callback.
- At that point, we may just be able to call `ValueBoundsConstraintSet::computeDependentBound`, which gives you a bound in which only certain SSA values (`vscale`) are allowed to appear.

Reply:
I don't think `ValueBoundsConstraintSet::computeDependentBound()` does exactly what I want here. First, I'm not sure I want to stop at the first `vscale` found, also `computeDependentBound()`  does not project out the `valueDim`, which in my experiments gave `affine_maps` take take two input symbols (presumably one for `vscale` and the `valueDim`), which did not usefully describe the bound.

Reply:
(I think the other points above could work :+1:)

Reply:
I looked into the above, and I don't really think it helps much with reuse, and is a little complex to setup. To be able to `dyn_cast` the `ValueBoundConstraintSet` we'd need to implement LLVM RTTI for it, and then it still only makes sense to use the helper on `ScalableValueBoundsConstraintSet` to compute the bound. The other variants construct a non-scalable `ValueBoundsConstraintSet` internally, so would not be useful for computing bounds dependent on `vscale`. 

And like I mentioned before none of the existing bounds functions do exactly what I want here, so I think I'd still need `computeScalableBound()` (and it's accompanying helpers). 

I think it could be done, but I think (right now) all it'd really eliminate is the `PopulateCustomValueBoundsFn` callback.

Reply:
I've now reworked things things as suggested above (which has removed the `PopulateCustomValueBoundsFn` callback).